### ISSUE TITLE
[Bug] gen_candidates_* fails with qKnowledgeGradient

### ISSUE THREAD

**Issue by wjmaddox** [pytorch/botorch#765]
# 🐛 Bug

There's a second backwards error that occurs when trying to use `gen_candidates_torch` with `qKnowledgeGradient`. I'm pretty confident that something isn't getting destroyed in the fantasy model after we evaluate it, but don't know what it is. `gen_candidates_torch` does work with other MC acquisition functions like `qExpectedImprovement`. 

## To reproduce

** Code snippet to reproduce **
```python
import torch

from botorch.models import SingleTaskGP
from botorch.generation.gen import gen_candidates_torch
from botorch.test_functions import Beale
from botorch.acquisition import qKnowledgeGradient, qExpectedImprovement
from botorch.models.transforms import Standardize
from botorch.optim.fit import fit_gpytorch_torch

from gpytorch.mlls import ExactMarginalLogLikelihood

torch.random.manual_seed(20)

neg_hartmann6 = Beale(negate=True)

bounds = neg_hartmann6.bounds

dtype = torch.double # if dtype=="double" else torch.float
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

def generate_initial_data(n, fn, NOISE_SE, device, dtype):
    # generate training data
    train_x = torch.rand(n, fn.dim, device=device, dtype=dtype)
    exact_obj = fn(train_x).unsqueeze(-1)  # add output dimension
    train_obj = exact_obj + NOISE_SE * torch.randn_like(exact_obj)

    train_obj = (train_obj - train_obj.mean() ) / train_obj.std()
    best_observed_value = train_obj.max().item()
    return train_x, train_obj, best_observed_value

# call helper functions to generate initial training data and initialize model
train_x_ei, train_obj_ei, best_observed_value_ei = generate_initial_data(
        n=50, fn=neg_hartmann6, NOISE_SE=0.2, device=device, dtype=dtype
)

model = SingleTaskGP(train_x_ei, train_obj_ei,)

bounds  = bounds.to(device)

xvals = torch.rand(5, 67, 2, device = device, dtype=dtype)
xvals =  bounds[0] + (bounds[1] - bounds[0]) * xvals

kg_exact = qKnowledgeGradient(exact, current_value=best_observed_value_ei, num_fantasies=64)
# the ei equivalent works
# kg_exact = qExpectedImprovement(exact, best_f=best_observed_value_ei)

gen_candidates_torch(xvals.detach(), kg_exact)
# this fails on iteration 2 with a double backward error
```

** Stack trace/error message **
```
RuntimeError: Trying to backward through the graph a second time, but the saved intermediate results have already been freed. Specify retain_graph=True when calling .backward() or autograd.grad() the first time.
```

## Expected Behavior

Pretty sure this should be okay.

## System information

**Please complete the following information:**
torch 1.8.0
gpytorch 1.4.0 (master)
botorch **'0.3.3.dev2+gdc868cc5.d20201108'** (master?)
mac os

## Additional context

Noticed this while trying to debug a research code related memory leak.

**Comment by Balandat** [comment]
Interesting. This works with `gen_candidates_scipy` - we should be able to traced down the differences in the gradient handling between the two.

**Comment by wjmaddox** [comment]
Ahh weird, it wasn't working with `gen_candidates_scipy` for me in my original repo; will take a closer look now.

**Comment by saitcakmak** [comment]
> I'm pretty confident that something isn't getting destroyed in the fantasy model after we evaluate it, but don't know what it is.

Yeah, that seems to be it. I stripped everything else from qKG and I still get the error. If I then get rid of the fantasy step, then the issue disappears. 

I think it is the `new_model.prediction_strategy.covar_cache` here https://github.com/cornellius-gp/gpytorch/blob/cc717f9fbddb1d4a4e868f2411a66502ce1b1a32/gpytorch/models/exact_gp.py#L230-L232 
If I replace that line with `new_model.prediction_strategy.covar_cache = None`, it works.

**Comment by wjmaddox** [comment]
In that case, then a fix is to do something like this here: https://github.com/pytorch/botorch/blob/8e357defb48ae7ebe09a547d14999e4de6ec416a/botorch/generation/gen.py#L224

```python
            bayes_optimizer.zero_grad()
            candidates.grad = None
            output_grad = torch.autograd.grad(loss, candidates)[0]
            candidates.grad = output_grad
```

**Comment by wjmaddox** [comment]
@saitcakmak were you setting the prediction strategy to non before or after the fantasy model call? 

Curious, because I think your solution has a better chance of solving my actual memory leak issue...

**Comment by saitcakmak** [comment]
> then a fix is to do something like this here
```
        def closure():
            bayes_optimizer.zero_grad()
            candidates.grad = None
            output_grad = torch.autograd.grad(loss, candidates)[0]
            candidates.grad = output_grad
            return loss
```
This doesn't throw an error either.

> were you setting the prediction strategy to non before or after the fantasy model 

I just changed this line
```
        new_model.prediction_strategy = old_pred_strat.get_fantasy_strategy(
            inputs, targets, full_inputs, full_targets, full_output, **fantasy_kwargs
        )
```
to `new_model.prediction_strategy = None`, effectively destroying the caches.

>  I think your solution has a better chance of solving my actual memory leak issue

Yeah, deleting the caches would probably help with that. I'd be worried about the computational cost of re-computing it with each fantasy call though

**Comment by Balandat** [comment]
Looks like the cause of the memory issues is https://github.com/cornellius-gp/gpytorch/issues/1560. We do need to fix the RuntimeError though, so leaving this open util #766 gets merged in.

**Comment by Balandat** [comment]
Resolved by #766
